### PR TITLE
A Flash Code Rework

### PR DESCRIPTION
**Pull request by BurgerLUA**
# Overview
I took one look at the flash code and was incredibly disappointed in how bad it was performance wise and beauty wise. It's incredibly old code and I genuinely believe that it needs an update.

## New flash design
There are three types of flashes:
Weak flashes, which have low battery and are burnt out easily.
Strong flashes, which have high battery and don't burn out quite as easily.
Cyborg flashes, which are like strong flashes however they cannot be modified.

Weak flashes are given to non-security personnel, like the Captain, CMO, and the roboticist for robotic eyes. Strong flashes are given to security personnel. Cyborg flashes are given to all cyborgs.

The flashes now run on power cells that can be swapped as long as you have a screwdriver on hand. As an added bonus, the bulb itself can be removed and replaced when burnt, so the whole assembly isn't useless. Cyborg flashes have this exception, and can be repaired by going into a cyborg charger just like it currently works.

## New flash mechanics
I've decided to only add confusion as an additional effect to flashes. I was going to add blindness and blurry vision, but currently the system is absolute garbage as the framework only runs every 2 seconds and it's absolutely jarrying.

### WIP
I still need to test the cyborg flashes. and I'm awaiting an master to dev merge to take place so shev's sprite screen flash duration changes are implemented.

### CODE REVIEW

**Comment by BotBOREALIS** [comment]
Icons with diff:
<details><summary>icons/obj/flash.dmi</summary>

Key | Old | New | Status
--- | --- | --- | ---
base||base|Created
bulb_burnt||bulb_burnt|Created
bulb_normal||bulb_normal|Created
bulb_weak||bulb_weak|Created
bulb_strong||bulb_strong|Created
flash_animation||flash_animation|Created
flash_animation[1]||![flash_animation[1]](http://images.aurorastation.org/images/63f42d44030c30493824e2d2668eb4bd.png)|Created
flash_animation[2]||![flash_animation[2]](http://images.aurorastation.org/images/6e59dae5d8715f9460c13532a4a48f2b.png)|Created
flash_animation[3]||![flash_animation[3]](http://images.aurorastation.org/images/bce1f2826cefb7720d65e9f8b91a69ae.png)|Created
flash_animation[4]||![flash_animation[4]](http://images.aurorastation.org/images/03d1632987502b1b4a07aa1ebf7f895f.png)|Created
flash_animation[5]||![flash_animation[5]](http://images.aurorastation.org/images/01c8e01b3000c52cc6ce3e20e918d594.png)|Created
flash_animation[6]||![flash_animation[6]](http://images.aurorastation.org/images/89d7c909e79b5fa248988b11ddaeafe8.png)|Created
flash_animation[7]||![flash_animation[7]](http://images.aurorastation.org/images/a0a3a311d4530468667b9140651a5acf.png)|Created
flash_animation[8]||![flash_animation[8]](http://images.aurorastation.org/images/61222fe7a21523bae3097920aa3941b1.png)|Created
flash_animation[9]||![flash_animation[9]](http://images.aurorastation.org/images/4f363636d57cc3ac0b5231d487e39ba5.png)|Created
flash_animation[10]||![flash_animation[10]](http://images.aurorastation.org/images/ab7efd473ace1a0396933a8682cc4b08.png)|Created
flash_animation[11]||![flash_animation[11]](http://images.aurorastation.org/images/52e87261c62f0001515e95a9e00c23a2.png)|Created
battery_hatch||battery_hatch|Created
battery_good||battery_good|Created
battery_low||battery_low|Created
battery_very_low||battery_very_low|Created
battery_none||battery_none|Created
battery_none[1]||![battery_none[1]](http://images.aurorastation.org/images/52e87261c62f0001515e95a9e00c23a2.png)|Created
</details>

**Comment by BurgerLUA** [comment]
!wip

**Comment by OneOneThreeEight** [comment]
👍 

looks fine now since my original issues were addressed

File: code/game/objects/items/devices/flash/core.dm [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	var/obj/item/weapon/cell/power_supply //The stored power cell. Internal value.
+	var/obj/item/weapon/flash_bulb/bulb //The installed bulb. Internal value.
+
+	var/self_recharge = FALSE
+	var/use_external_power = FALSE
+	var/charge_cost = 100 //How much charge to take away per use
+
+	var/recharge_time = 4//No clue
+
+	var/DRM = FALSE //True means you can't modify it.
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: Not a great variable name. Should be lower-case and actually describe what it does.

File: code/game/objects/items/devices/flash/core.dm [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+	if(bulb)
+		bulb.on = TRUE
+
+	update_icon()
+
+/obj/item/device/flash/examine(mob/user)
+	..(user)
+	var/uses_remaining = round(power_supply.charge / charge_cost)
+	to_chat(user,"Has [uses_remaining] use\s remaining.")
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: Check the parent call's return value, you shouldn't be able to check this from across the room.

File: code/game/objects/items/devices/flash/core.dm [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+/obj/item/device/flash/attack(mob/living/M as mob, mob/living/user as mob, var/target_zone)
+
+	user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+	user.do_attack_animation(M)
+
+	if(strength_mul <= 0 || !power_supply || !power_supply.checked_use(charge_cost) || !bulb || bulb.is_burnt)
+		playsound(src.loc, 'sound/items/flashlight.ogg', 50, 1)
+		if(M != user)
+			user.visible_message("<span class='notice'>\The [user] sticks \the [src] in \the [src]'s direction, but nothing happens.</span>", \
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: Awkward phrasing.

File: code/game/objects/items/devices/flash/core.dm [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+	addtimer(CALLBACK(src, .proc/try_recharge), recharge_time * 2 SECONDS, TIMER_UNIQUE)
+
+/obj/item/device/flash/attack(mob/living/M as mob, mob/living/user as mob, var/target_zone)
+
+	user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
+	user.do_attack_animation(M)
+
+	if(strength_mul <= 0 || !power_supply || !power_supply.checked_use(charge_cost) || !bulb || bulb.is_burnt)
+		playsound(src.loc, 'sound/items/flashlight.ogg', 50, 1)
+		if(M != user)
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: You do understand how `visible_message()` works, yes?

File: code/game/objects/items/devices/flash/core.dm [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+				"<span class='notice'>You hear a click.</span>")
+		return
+
+	playsound(src.loc, 'sound/weapons/flash.ogg', 50, 1)
+
+	src.update_icon()
+	src.add_overlay("flash_animation")
+
+	if(M == user) //Radial flash
+		user.visible_message("<span class='warning'>\The [user.name] sticks \the [src] in the air and presses the activation button!</span>", \
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: Awkward phrasing. Surely there's a better term to use than "sticks".

File: code/game/objects/items/devices/flash/core.dm [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+			"<span class='warning'>You stick \the [src] in the air and press the activation button!</span>", \
+			"<span class='notice'>You hear a distinct mechanical shutter...</span>")
+		for (var/mob/living/T in view(user))
+			if(T == user)
+				continue
+			var/distance_mod = 1 - min(1,get_dist(user,T)/6)
+			var/total_power = bulb.strength*strength_mul*0.75*distance_mod
+			if(total_power > 0)
+				T.visible_message("<span class='danger'>\The [T] is caught in the flash!</span>", \
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: Why exactly is this a visible_message? This'll result in a lot of chat spam if there's a lot of people in range.

File: code/game/objects/items/devices/flash/core.dm [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+	if (self_recharge)
+		addtimer(CALLBACK(src, .proc/try_recharge), recharge_time * 2 SECONDS, TIMER_UNIQUE)
+
+	bulb.add_heat(strength_mul)
+
+/obj/item/device/flash/attackby(obj/item/weapon/W as obj, mob/user as mob)
+
+	var/obj/item/weapon/cell/the_cell = W
+	var/obj/item/weapon/flash_bulb/the_bulb = W
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/core.dm]: ...why preemptively cast both types like this?

File: code/game/objects/items/devices/flash/effects.dm [hunk 8]
```diff
@@ -0,0 +1,57 @@
+//Most of the things like charging and power supply were stolen from energy gun code.
+/mob/proc/on_flash(var/mob/living/attacker as mob, var/obj/weapon, var/severity = 1)
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/effects.dm]: `as mob` is meaningless in non-verbs like this.

File: code/game/objects/items/devices/flash/effects.dm [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	//It is possible that can be no attacker (See: mounted flash)
+	//Generate logs
+	if(attacker.ckey && src.ckey) //Only give a shit if both the attacker and the src have a ckey
+		src.attack_log += text("\[[time_stamp()]\] <font color='orange'>Has been flashblinded with \the [weapon.name]  by [attacker.name] ([attacker.ckey]) with a severity of [severity].</font>")
```
Comment (Lohikar) [inline code/game/objects/items/devices/flash/effects.dm]: There are procs for logging attacks, please use them.

File: code/game/objects/items/robot/robot_parts.dm [hunk 10]
```diff
@@ -304,7 +304,17 @@
 			user << "<span class='notice'>You enable the lawing circuits on \the [src].</span>"
 			law_manager = TRUE
 
-	if(istype(W, /obj/item/device/flash))
+	var/obj/item/device/flash/selected_flash = W
```
Comment (Lohikar) [inline code/game/objects/items/robot/robot_parts.dm]: Why preemptively cast?

File: code/modules/mob/living/silicon/robot/robot_modules.dm [hunk 11]
```diff
@@ -112,14 +112,14 @@ var/global/list/robot_modules = list(
...
 	var/obj/item/device/flash/F = locate() in src.modules
 	if(F)
-		if(F.broken)
-			F.broken = 0
-			F.times_used = 0
-			F.icon_state = "flash"
-		else if(F.times_used)
-			F.times_used--
+		var/obj/item/weapon/flash_bulb/bulb = F.bulb
+		if(bulb && istype(bulb))
```
Comment (Lohikar) [inline code/modules/mob/living/silicon/robot/robot_modules.dm]: Nullcheck on bulb is redundant; `istype(null)` is always `FALSE`.

**Comment by BurgerLUA** [comment]
Tomorrow I'll figure out the attack log proc as well, and fix that one missing as mob cast that I forgot to change.

**Comment by BurgerLUA** [comment]
This is almost done. I just got to see if mounted flashes still work with cyborgs and rigs, and it's gucci.

**Comment by BurgerLUA** [comment]
Unless Travis "Three Reichs and you're out" CI has any objections, this is done.

**Comment by BurgerLUA** [comment]
!review

**Comment by BurgerLUA** [comment]
I tried updating the branch but it caused some errors.

File: code/game/objects/items/devices/flash/effects.dm [hunk 12]
```diff
@@ -0,0 +1,58 @@
+//Most of the things like charging and power supply were stolen from energy gun code.
+/mob/proc/on_flash(var/mob/living/attacker, var/obj/weapon, var/severity = 1)
+	//It is possible that can be no attacker (See: mounted flash)
+	//Generate logs
+	if(attacker.ckey && src.ckey) //Only give a shit if both the attacker and the src have a ckey
+		admin_attack_log(attacker, src, "used a flash", "Was subjected to a flash", "used a flash on")
+
+	return severity
+
+/mob/living/carbon/human/on_flash(var/mob/living/attacker, var/obj/weapon, var/severity = 1)
```
Comment (Sindorman) [inline code/game/objects/items/devices/flash/effects.dm]: This includes IPCs, and it seems like you plan IPCs to be affected by flash?

File: code/game/objects/items/devices/flash/effects.dm [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+		src.druggy = max(src.druggy,(60/2) * severity)
+		var/obj/item/organ/eyes/E = src.get_eyes()
+		if(E && !(E.status & ORGAN_ROBOT))
+			to_chat(src,"<span class='warning'>Your eyes burn with the intense light of the flash!</span>")
+			E.take_damage(5 * severity)
+		src.Weaken(10 * severity)
+
+	return severity
+
+/mob/living/carbon/human/diona/on_flash(var/mob/living/attacker, var/obj/weapon, var/severity = 1)
```
Comment (skull132) [inline code/game/objects/items/devices/flash/effects.dm]: This class cannot be used. Subclasses of human are spawn helpers and should _not_ be evolved into logical components. Species specific logic has to reside in the species datum.

File: html/changelogs/burgerbb-flash_rework.yml [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+# Any changes you've made.  See valid prefix list above.
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, all entries are changed into a single [] after a master changelog generation. Just remove the brackets when you add new entries.
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - rscadd: "Flashes are now modular. They can take any sort of power cell, and the bulb can be replaced. Weak flash bulbs can be produced as autolathes, normal ones are orderable from cargo."
+  - tweak: "Tweaked how flash activation works. Self-targeting yourself with a flash will to a radial blind instead of flashing yourself."
+  - balance: "Increased the cooldown of rig mounted flashes to 2 seconds, and made mounted flash bulbs unable to burn out."
+  - balance: "Flashes now cause confusion (random movement stepping) when used on organics. Flashes cause 'druggy' efects on vaurca."
```
Comment (skull132) [inline html/changelogs/burgerbb-flash_rework.yml]: Effects*

File: html/changelogs/burgerbb-flash_rework.yml [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+delete-after: True
+
+# Any changes you've made.  See valid prefix list above.
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, all entries are changed into a single [] after a master changelog generation. Just remove the brackets when you add new entries.
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - rscadd: "Flashes are now modular. They can take any sort of power cell, and the bulb can be replaced. Weak flash bulbs can be produced as autolathes, normal ones are orderable from cargo."
+  - tweak: "Tweaked how flash activation works. Self-targeting yourself with a flash will to a radial blind instead of flashing yourself."
```
Comment (skull132) [inline html/changelogs/burgerbb-flash_rework.yml]: Self-targeting already implies targeting yourself. 🤔

File: html/changelogs/burgerbb-flash_rework.yml [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+# Optional: Remove this file after generating master changelog.  Useful for PR changelogs that won't get used again.
+delete-after: True
+
+# Any changes you've made.  See valid prefix list above.
+# INDENT WITH TWO SPACES.  NOT TABS.  SPACES.
+# SCREW THIS UP AND IT WON'T WORK.
+# Also, all entries are changed into a single [] after a master changelog generation. Just remove the brackets when you add new entries.
+# Please surround your changes in  double quotes ("), as certain characters otherwise screws up compiling. The quotes will not show up in the changelog.
+changes: 
+  - rscadd: "Flashes are now modular. They can take any sort of power cell, and the bulb can be replaced. Weak flash bulbs can be produced as autolathes, normal ones are orderable from cargo."
```
Comment (skull132) [inline html/changelogs/burgerbb-flash_rework.yml]: produced at*
(can be ordered from*)

**Comment by skull132** [comment]
To clearly outline what needs to happen:

 - The bulb mechanic can go, it's needless complexity in a place that should be simple. A flash and a synth flash is good enough. Key point: flashes are simple, common devices. Their usage and (lack of) variety should represent that.
 - The battery shenanigans can stay.
 - Keep the weaken mechanics. We can trial them.

**Pull request closed without merging**